### PR TITLE
Feat/add feesmanager contract

### PR DESCRIPTION
**Pull request by envin3**
Add fees manager contract

### CODE REVIEW

File: cpp/contracts/Makefile [hunk 1]
```diff
@@ -1,10 +1,13 @@
...
-%.wasm: %.cpp
+build/%.wasm: %.cpp | build
 	cdt-cpp -o $@ $<
 
+build:
+	mkdir -p build
+
 clean:
-	rm -f *.abi *.wasm
+	rm -f ./build/*.abi ./build/*.wasm
```
Comment (gitmp01) [inline cpp/contracts/Makefile:13]: N.I.C.E

File: cpp/contracts/feesmanager.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
+#include "feesmanager.hpp"
+
+using namespace eosio;
+
+name owner;
+
+void feesmanager::init( name security_council ) {
```
Comment (gitmp01) [inline cpp/contracts/feesmanager.cpp:7]: Just wondering, is this space before/after parenthesis in the function prototype a thing in the usual cpp style guide? Because it makes my eyes bleeding... just asking :sweat_smile:

File: cpp/contracts/feesmanager.hpp [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+private:
+    TABLE allowance {
+        name node;
+        name token;
+        uint64_t amount;
+
+        uint64_t primary_key() const { return node.value + token.value; }
+        
+        uint64_t by_token() const { return token.value; }
+    };
```
Comment (gitmp01) [inline cpp/contracts/feesmanager.hpp:31]: Thinking about it this is not the "eos" way of doing this. If you check the `accounts` table in the `eosio.token.hpp` contract, you'll see this table:

```cpp
struct [[eosio::table]] account {
   asset    balance;

   uint64_t primary_key()const { return balance.symbol.code().raw(); }
};
```
Where you can access it like this:

```cpp
accounts from_acnts( get_self(), owner.value );
```

You can see the scope is restricted to `owner.value`, meaning there will be a different table for each account that will have some allowance. And it is indexed by the token symbol code, so each table per account will have a set of symbols with the specific balance.

Also, since EOS works with symbols rather then accounts for identifying tokens, I would use the `asset` type as done for the `balance` column in the `accounts` table.

**tl;dr** just copy the table above from `eosio.token.hpp` and rename `balance` to `allowance`

**Review by gitmp01 — CHANGES_REQUESTED** [review]
See the comment related to the table, let me know if you have any opinions, but I think there is the chance that the audit will make us change it if we are not going to do it now.

File: cpp/contracts/feesmanager.cpp [hunk 4]
```diff
@@ -12,58 +12,56 @@ void feesmanager::init( name security_council ) {
...
         "transfer"_n,
-        std::make_tuple(get_self(), node, allowance_itr->amount, std::string("Withdraw"))
+        std::make_tuple(get_self(), node, asset_data, std::string("Withdraw"))
     ).send();
 }
 
-void feesmanager::withdrawto(name node, const std::vector<name>& tokens) {
+void feesmanager::withdrawto(name node, const std::vector<asset>& tokens) {
     for (const auto& token : tokens) {
         withdrawto(node, token);
```
Comment (gitmp01) [inline cpp/contracts/feesmanager.cpp:66]: Double check this is working in a test because I'm not 100% sure it will, since `withdrawto` is declared as an action. 

If it doesn't you'll probably need to factor out the code within `withdrawto` to an "internal" function (which is not declared as ACTION) that can be called as needed.

**Review by gitmp01 — CHANGES_REQUESTED** [review]
Looks good! I would just add a test unit testing the owner-methods will revert if called by someone else

File: cpp/contracts/feesmanager.cpp [hunk 5]
```diff
@@ -0,0 +1,76 @@
+#include "feesmanager.hpp"
+
+using namespace eosio;
+
+void feesmanager::setallowance( name node, name token, const asset& value ) {
+    require_auth(get_self());
+
+    auto balance = getbalance(token, value.symbol);
+    check(balance.amount >= value.amount, "balance is lower than the allowance to be set");
```
Comment (gitmp01) [inline cpp/contracts/feesmanager.cpp:9]: uhm don't you need to consider the allowance that has allocated already?
```cpp
// pseudo
setallowance(node, token, value) {
  b = getBalance(feesmanager) 
  x = currentNodeAllowance(node)
  t = getTotalAllocatedAllowance() // you need a singleton tracking this I think
  check(y < b - (t - x), "too much bro!")
  // ...
}
```

File: cpp/contracts/feesmanager.cpp [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+void feesmanager::incallowance( name node, name token, const asset& value ) {
+    require_auth(get_self());
+
+    allowances_table allowances(get_self(), node.value);
+    const auto& allowance_table = allowances.get(value.symbol.code().raw(), "No allowance set for this node");
+
+    auto balance = getbalance(token, value.symbol);
+    check(balance.amount >= value.amount + allowance_table.allowance_data.amount, "balance is lower than the allowance to be set");
```
Comment (gitmp01) [inline cpp/contracts/feesmanager.cpp:36]: A similar check to the above needs to be applied here:
```cpp
incallowance(node, token, value) {
  // ...
  check(value < b - t, "too much bro!")
  // ...
}
```

**Review by gitmp01 — CHANGES_REQUESTED** [review]
I think there are problems, see inline comments

File: cpp/contracts/feesmanager.cpp [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    if (allowance_itr == allowances.end() && allowance_name_itr == allowance_name_idx.end()) {
+        check(balance.amount >= total_token_allowance + value.amount, "balance is lower than the allowance to be set");
+        allowances.emplace(get_self(), [&](auto& a) {
+            a.node_allowance = value;
+            a.token = token;
+        });
+        total_allowance.set({asset(total_token_allowance + value.amount , value.symbol)}, get_self());
+    } else {
+        check(allowance_itr->token == token, "symbol and token do not match");
+        check(balance.amount >= (total_token_allowance - allowance_itr->node_allowance.amount) + value.amount, "balance is lower than the allowance to be set");
```
Comment (gitmp01) [inline cpp/contracts/feesmanager.cpp:27]: I'd differentiate this check msg from the previous one, so we can know on which branch of the if we are

File: cpp/test/feesmanager.test.js [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+      const balance = token.contract.tables.accounts(getAccountCodeRaw(feesmanager.account)).getTableRow(assetSymbol)
+      expect(balance).to.be.deep.equal({
+        balance: feesmanagerBalance,
+      })
+
+      const action = feesmanager.contract.actions
+        .setallowance([node1, token.account, allowanceValue])
+        .send(active(feesmanager.account))
+
+      await expectToThrow(action, errors.INSUFFICIENT_BALANCE)
```
Comment (gitmp01) [inline cpp/test/feesmanager.test.js:95]: good!

**Review by gitmp01 — APPROVED** [review]
LGMT, good job!
